src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp

```diff
@@ -473,7 +474,156 @@ pass::EliminateSqueeze::EliminateSqueeze() {
+        // for Concat and the last input to Concat is output(1) of Sequence op, which is actually Hn.
+        // This is also a valid case for this Elimination.
+        if (!cast_to_split) {
+            if (idx != (concat_in_values.size() - 1) || !split) {
+                return {};
+            }
+            shared_ptr<Node> in_to_split = split->input_value(0).get_node_shared_ptr();
+            Output<Node> seq_out;
+            if (in_to_split && !in_to_split->inputs().empty()) {
+                seq_out = in_to_split->input_value(0);
+            } else {
+                return {};
+            }
+
+            auto seq_node = seq_out.get_node_shared_ptr();
+            if (!seq_node || seq_out.get_index() != 0 ||
+                !(dynamic_pointer_cast<ov::opset9::RNNSequence>(seq_node) ||
+                  dynamic_pointer_cast<ov::opset9::GRUSequence>(seq_node) ||
+                  dynamic_pointer_cast<ov::opset9::LSTMSequence>(seq_node))) {
+                return {};
+            }
+
+            // check that Split is connected to Sequence->output(0)
+            // possible patterns:
+            // Sequence:0->Squeeze->Split
+            bool valid_pattern = check_squeeze(in_to_split);
+            // Sequence:0->Reshape->Split
+            if (!valid_pattern) {
+                valid_pattern = check_reshape(in_to_split);
+            }
+
+            if (!valid_pattern) {
+                return {};
+            }
+
+            // check that Sequence->output(1) is connected to this input
+            if (!seq_node || in_to_concat != seq_node->output(1)) {
+                return {};
+            }
+            return split;
+        }
+        // input (split op) should be the same for all inputs
+        if (!split) {
+            split = cast_to_split;
+        } else if (cast_to_split.get() != split.get()) {
+            // not all inputs to concat belong to the same Split op
+            return {};
+        }
+
+        // Split to Concat edges are not in orderl
+        // should be (0, 1, 2, ... , split->outputs().size()-1)
+        if (in_to_concat.get_index() != idx) {
+            return {};
+        }
+        ++idx;
+    }
+
+    // not all split outputs are used.
+    if (idx != split->outputs().size()) {
+        return {};
+    }
+
+    return split;
+}
+}  // namespace
+
+ov::pass::EliminateSplitConcat::EliminateSplitConcat() {
+    MATCHER_SCOPE(EliminateSplitConcat);
+
+    auto pattern_concat = pattern::wrap_type<opset8::Concat>();
+    matcher_pass_callback callback = [=](pattern::Matcher& m) {
+        const auto& pattern_map = m.get_pattern_map();
+        const auto concat = dynamic_pointer_cast<ov::opset9::Concat>(pattern_map.at(pattern_concat));
+        if (!concat) {
+            return false;
+        }
+        shared_ptr<Node> split = check_all_inputs<ov::opset9::Split>(concat);
+        if (!split) {
+            split = check_all_inputs<ov::opset9::VariadicSplit>(concat);
```

Comment:
Variadic Split could have divided the data into uneven chunks -- the other way then they were concatenated. Especially this could be dangerous if the logic with Sequence is engaged. If we used Variadic Split followed by Concat to cut more than we concatenate from Sequence. The same question goes to Split with num_splits != Hn

Reply:
it looks like it's enough to sheck that` last_variadic_split_output_shape[split_axis] == 1` that has been implemented

Reply:
Looks like it but I don't see this check in the code. Am I missing smth?

Reply:
fixed